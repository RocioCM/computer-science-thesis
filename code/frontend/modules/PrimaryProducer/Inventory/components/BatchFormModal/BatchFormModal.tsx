import Modal, { ModalProps } from '@/common/components/Modal';
import { BottleBatch } from '../../types';
import PrimaryProducerServices from '../../services';
import { toast } from 'react-toastify';
import useForm from '@/common/hooks/useForm';
import {
  BATCH_COMPOSITION_FORM_INPUTS,
  BATCH_FORM_INPUTS_1,
  BATCH_FORM_INPUTS_2,
  BATCH_FORM_STRUCT,
} from '../../constants';
import Button from '@/common/components/Button';
import { useEffect } from 'react';
import FaIcon from '@/common/components/FaIcon';

export interface Props extends ModalProps {
  handleCancel: () => any;
  editingId: number | null;
  handleSuccess: () => void;
}

const BatchFormModal = ({
  editingId,
  handleCancel,
  handleSuccess,
  ...props
}: Props) => {
  const { form, handleChange, formBuilder, submitEnabled, setInitialForm } =
    useForm(BATCH_FORM_STRUCT);
  const isEditing = !!editingId;

  const fetchBatchData = async () => {
    if (!editingId) return;
    const { ok, data } = await PrimaryProducerServices.getBatch(editingId);
    if (ok) {
      setInitialForm(data);
    } else {
      toast.error('Ocurri贸 un error al cargar los datos del lote de botellas');
    }
  };

  const createBottleBatch = async (data: Partial<BottleBatch>) => {
    const payload = data;
    const { ok } = await PrimaryProducerServices.createBatch(payload);
    if (ok) {
      toast.success('Lote de botellas creado correctamente');
      handleCancel();
      handleSuccess();
    } else {
      toast.error('Ocurri贸 un error al crear el lote de botellas');
    }
  };

  const updateBottleBatch = async (data: Partial<BottleBatch>) => {
    if (!editingId) return;
    const id = editingId;
    const payload = { ...data, id };
    const { ok } = await PrimaryProducerServices.updateBatch(payload);
    if (ok) {
      toast.success('Lote de botellas actualizado correctamente');
      handleCancel();
      handleSuccess();
    } else {
      toast.error('Ocurri贸 un error al actualizar el lote de botellas');
    }
  };

  const handleSubmit = async () => {
    if (editingId) {
      await updateBottleBatch(form);
    } else {
      await createBottleBatch(form);
    }
  };

  const handleAddCompositionItem = () => {
    const composition = form.bottleType?.composition
      ? [...form.bottleType.composition]
      : [];
    composition.push({ name: '', amount: '', measureUnit: '' });
    handleChange('bottleType.composition', composition);
  };

  const handleDeleteCompositionItem = (index: number) => {
    const composition = form.bottleType?.composition
      ? [...form.bottleType.composition]
      : [];
    composition.splice(index, 1);
    handleChange('bottleType.composition', composition);
  };

  useEffect(() => {
    if (editingId) fetchBatchData();
  }, [editingId]);

  return (
    <Modal handleCancel={handleCancel} big skippable={false} {...props}>
      <h2 className="w-full bg-n0">
        {isEditing ? 'Editar' : 'Crear'} lote de botellas{' '}
        {editingId ? `#${editingId}` : ''}
      </h2>
      <form className="h-full w-full overflow-auto hide-scroll relative pt-2">
        <div className="flex flex-col gap-2">
          <div className="grid grid-cols-2 gap-2">
            {formBuilder(BATCH_FORM_INPUTS_1)}
          </div>
          <hr className="my-4" />
          <div className="flex flex-col gap-2">
            <h3>Composici贸n</h3>
            {form.bottleType?.composition?.map((_: any, i: number) => (
              <div key={i} className="flex gap-2">
                {formBuilder(
                  i === 0
                    ? BATCH_COMPOSITION_FORM_INPUTS
                    : BATCH_COMPOSITION_FORM_INPUTS.map((input) => ({
                        ...input,
                        label: '',
                      })),
                  {
                    prefix: `bottleType.composition.${i}`,
                  }
                )}
                {form.bottleType.composition.length > 1 && (
                  <Button
                    title="Eliminar"
                    variant="secondary"
                    handleClick={() => handleDeleteCompositionItem(i)}
                    className={i === 0 ? 'mt-[1.8rem]' : ''}
                  >
                    <FaIcon type="fa-solid fa-trash-can" />
                  </Button>
                )}
              </div>
            ))}
            <Button variant="secondary" handleClick={handleAddCompositionItem}>
              <FaIcon type="fa-solid fa-plus" /> Agregar material
            </Button>
          </div>
          <hr className="my-4" />
          <div className="grid grid-cols-2 gap-2">
            {formBuilder(BATCH_FORM_INPUTS_2)}
          </div>
        </div>
        <Button
          label={isEditing ? 'Actualizar' : 'Crear'}
          disabled={!submitEnabled}
          className="ml-auto mt-4 sticky bottom-0 min-w-[10rem]"
          handleClick={handleSubmit}
        />
      </form>
    </Modal>
  );
};

export default BatchFormModal;
