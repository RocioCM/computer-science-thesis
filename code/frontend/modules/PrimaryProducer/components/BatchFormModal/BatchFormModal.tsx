import Modal, { ModalProps } from '@/common/components/Modal';
import { BottleBatch } from '../../types';
import PrimaryProducerServices from '../../services';
import { toast } from 'react-toastify';
import useForm from '@/common/hooks/useForm';
import {
  BATCH_COMPOSITION_FORM_INPUTS,
  BATCH_FORM_INPUTS_1,
  BATCH_FORM_INPUTS_2,
  BATCH_FORM_STRUCT,
} from '../../constants';
import Button from '@/common/components/Button';
import { useEffect } from 'react';

export interface Props extends ModalProps {
  handleCancel: () => any;
  editingId: number | null;
  handleSuccess: () => void;
}

const BatchFormModal = ({
  editingId,
  handleCancel,
  handleSuccess,
  ...props
}: Props) => {
  const { form, handleChange, formBuilder, submitEnabled, setInitialForm } =
    useForm(BATCH_FORM_STRUCT);
  const isEditing = !!editingId;

  const fetchBatchData = async () => {
    if (!editingId) return;
    const { ok, data } = await PrimaryProducerServices.getBatch(editingId);
    if (ok) {
      setInitialForm(data);
    } else {
      toast.error('Ocurri贸 un error al cargar los datos del lote de botellas');
    }
  };

  const createBottleBatch = async (data: Partial<BottleBatch>) => {
    const payload = data;
    const { ok } = await PrimaryProducerServices.createBatch(payload);
    if (ok) {
      toast.success('Lote de botellas creado correctamente');
      handleCancel();
      handleSuccess();
    } else {
      toast.error('Ocurri贸 un error al crear el lote de botellas');
    }
  };

  const updateBottleBatch = async (data: Partial<BottleBatch>) => {
    if (!editingId) return;
    const id = editingId;
    const payload = { ...data, id };
    const { ok } = await PrimaryProducerServices.updateBatch(id, payload);
    if (ok) {
      toast.success('Lote de botellas actualizado correctamente');
      handleCancel();
      handleSuccess();
    } else {
      toast.error('Ocurri贸 un error al actualizar el lote de botellas');
    }
  };

  const handleSubmit = () => {
    if (editingId) {
      updateBottleBatch(form);
    } else {
      createBottleBatch(form);
    }
  };

  const handleAddCompositionItem = () => {
    const composition = form.bottleType?.composition
      ? [...form.bottleType.composition]
      : [];
    composition.push({ name: '', amount: '', measureUnit: '' });
    handleChange('bottleType.composition', composition);
  };

  const handleDeleteCompositionItem = (index: number) => {
    const composition = form.bottleType?.composition
      ? [...form.bottleType.composition]
      : [];
    composition.splice(index, 1);
    handleChange('bottleType.composition', composition);
  };

  useEffect(() => {
    if (editingId) fetchBatchData();
  }, [editingId]);

  return (
    <Modal handleCancel={handleCancel} big skippable={false} {...props}>
      <form className="h-full w-full overflow-auto relative">
        <h2 className="bg-n0 sticky top-0 pb-2">
          {isEditing ? 'Editar' : 'Crear'} lote de botellas
        </h2>
        <div className="flex flex-col gap-2">
          {formBuilder(BATCH_FORM_INPUTS_1)}
          <div className="flex flex-col gap-2">
            <h3>Composici贸n</h3>
            {form.bottleType?.composition?.map((_: any, i: number) => (
              <div key={i} className="flex gap-2">
                {formBuilder(BATCH_COMPOSITION_FORM_INPUTS, {
                  prefix: `bottleType.composition.${i}`,
                })}
                {i > 0 && (
                  <Button
                    label="Eliminar"
                    handleClick={() => handleDeleteCompositionItem(i)}
                  />
                )}
              </div>
            ))}
            <Button
              label="Agregar material"
              handleClick={handleAddCompositionItem}
            />
          </div>
          {formBuilder(BATCH_FORM_INPUTS_2)}
        </div>
        <Button
          label={isEditing ? 'Actualizar' : 'Crear'}
          disabled={!submitEnabled}
          className="ml-auto mt-4 sticky bottom-0 min-w-[10rem]"
          handleClick={handleSubmit}
        />
      </form>
    </Modal>
  );
};

export default BatchFormModal;
