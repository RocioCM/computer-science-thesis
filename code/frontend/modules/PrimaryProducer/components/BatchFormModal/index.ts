import BatchFormModal from './BatchFormModal';

export type { Props as BatchFormModalProps } from './BatchFormModal';
export default BatchFormModal;
