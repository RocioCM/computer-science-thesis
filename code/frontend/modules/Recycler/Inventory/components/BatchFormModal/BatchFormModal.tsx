import { useEffect, useState } from 'react';
import { toast } from 'react-toastify';
import Modal, { ModalProps } from '@/common/components/Modal';
import useForm from '@/common/hooks/useForm';
import { RecyclingBatch, WasteBottle } from '../../types';
import RecyclerServices from '../../services';
import {
  BATCH_COMPOSITION_FORM_INPUTS,
  BATCH_FORM_INPUTS_1,
  BATCH_FORM_STRUCT,
} from '../../constants';
import Button from '@/common/components/Button';
import FaIcon from '@/common/components/FaIcon';
import InputCheckbox from '@/common/components/Inputs/InputCheckbox';
import LoadingSpinner from '@/common/components/LoadingSpinner';

export interface Props extends ModalProps {
  handleCancel: () => any;
  editingId: number | null;
  handleSuccess: () => void;
}

const BatchFormModal = ({
  editingId,
  handleCancel,
  handleSuccess,
  ...props
}: Props) => {
  const [availableWasteBottles, setAvailableWasteBottles] = useState<
    WasteBottle[]
  >([]);
  const [loading, setLoading] = useState(false);
  const { form, handleChange, formBuilder, submitEnabled, setInitialForm } =
    useForm(BATCH_FORM_STRUCT);
  const isEditing = !!editingId;

  const fetchAvailableWasteBottles = async () => {
    setLoading(true);
    const { ok, data } = await RecyclerServices.getAvailableWasteBottles(
      1,
      100
    ); /// TODO: add pagination here.
    setLoading(false);
    if (ok) {
      setAvailableWasteBottles((prevBottles) => {
        const bottlesObj: Record<number, WasteBottle> = prevBottles.reduce(
          (acc, bottle) => ({ ...acc, [bottle.id]: bottle }),
          {}
        );
        data.forEach((bottle) => {
          bottlesObj[bottle.id] = bottle;
        });
        return Object.values(bottlesObj);
      });
    } else {
      toast.error('Ocurrió un error al cargar las botellas disponibles');
    }
  };

  const fetchBatchData = async () => {
    if (!editingId) return;
    const { ok, data } = await RecyclerServices.getRecyclingBatch(editingId);
    if (ok) {
      const formattedForm = {
        ...data,
        createdAt: data.createdAt
          ? new Date(data.createdAt).toISOString().split('T')[0]
          : null,
        wasteBottleIds: data.wasteBottleIds.reduce(
          (acc, id) => ({ ...acc, [id]: true }),
          {}
        ),
      };
      setInitialForm(formattedForm);
      setAvailableWasteBottles((prev) => [
        ...data.wasteBottleIds.map((id) => ({ id } as WasteBottle)),
        ...prev,
      ]);
    } else {
      toast.error('Ocurrió un error al cargar los datos del lote de material');
    }
  };

  const createBottleBatch = async (data: Partial<RecyclingBatch>) => {
    const payload = {
      ...data,
      wasteBottleIds: Object.entries(form.wasteBottleIds)
        .filter(([, value]) => value)
        .map(([key]) => Number(key)),
    };
    const { ok } = await RecyclerServices.createRecyclingBatch(payload);
    if (ok) {
      toast.success('Lote de material creado correctamente');
      handleCancel();
      handleSuccess();
    } else {
      toast.error(
        'Ocurrió un error al crear el lote de material. Intenta nuevamente más tarde'
      );
    }
  };

  const updateBottleBatch = async (data: Partial<RecyclingBatch>) => {
    if (!editingId) return;
    const id = editingId;
    const payload = {
      ...data,
      id,
      wasteBottleIds: Object.entries(form.wasteBottleIds)
        .filter(([, value]) => value)
        .map(([key]) => Number(key)),
    };
    const { ok } = await RecyclerServices.updateRecyclingBatch(payload);
    if (ok) {
      toast.success('Lote de material actualizado correctamente');
      handleCancel();
      handleSuccess();
    } else {
      toast.error('Ocurrió un error al actualizar el lote de material');
    }
  };

  const handleSubmit = async () => {
    if (editingId) {
      await updateBottleBatch(form);
    } else {
      await createBottleBatch(form);
    }
  };

  const handleAddCompositionItem = () => {
    const composition = form.composition ? [...form.composition] : [];
    composition.push({ name: '', amount: '', measureUnit: '' });
    handleChange('composition', composition);
  };

  const handleDeleteCompositionItem = (index: number) => {
    const composition = form.composition ? [...form.composition] : [];
    composition.splice(index, 1);
    handleChange('composition', composition);
  };

  const handleSelectAllBottles = () => {
    const isAlreadyAllSelected =
      Object.keys(form.wasteBottleIds).length === availableWasteBottles.length;
    if (isAlreadyAllSelected) {
      handleChange('wasteBottleIds', {});
    } else {
      handleChange(
        'wasteBottleIds',
        availableWasteBottles.reduce(
          (acc, bottle) => ({ ...acc, [bottle.id]: true }),
          {}
        )
      );
    }
  };

  useEffect(() => {
    if (editingId) fetchBatchData();
  }, [editingId]);

  useEffect(() => {
    fetchAvailableWasteBottles();
  }, []);

  return (
    <Modal handleCancel={handleCancel} big skippable={false} {...props}>
      <h2 className="w-full font-medium text-[1.75rem] bg-n0 border-b border-n2 pb-2 mb-1">
        {isEditing ? 'Editar' : 'Crear'} lote de material{' '}
        {editingId ? `#${editingId}` : ''}
      </h2>
      <form className="h-full w-full overflow-auto hide-scroll relative ">
        <div className="flex flex-col gap-2">
          <div className="grid grid-cols-2 gap-2">
            {formBuilder(BATCH_FORM_INPUTS_1)}
          </div>
          <hr className="my-4" />
          <div className="flex flex-col gap-2">
            <h3>Composición</h3>
            {form.composition?.map((_: any, i: number) => (
              <div key={i} className="flex gap-2">
                {formBuilder(
                  i === 0
                    ? BATCH_COMPOSITION_FORM_INPUTS
                    : BATCH_COMPOSITION_FORM_INPUTS.map((input) => ({
                        ...input,
                        label: '',
                      })),
                  {
                    prefix: `composition.${i}`,
                  }
                )}
                {form.composition.length > 1 && (
                  <Button
                    title="Eliminar"
                    variant="secondary"
                    handleClick={() => handleDeleteCompositionItem(i)}
                    className={i === 0 ? 'mt-[1.8rem]' : ''}
                  >
                    <FaIcon type="fa-solid fa-trash-can" />
                  </Button>
                )}
              </div>
            ))}
            <Button variant="secondary" handleClick={handleAddCompositionItem}>
              <FaIcon type="fa-solid fa-plus" /> Agregar material
            </Button>
          </div>
          <hr className="my-4" />
          <div className="flex flex-col gap-2">
            <h3>Botellas</h3>
            {loading ? (
              <LoadingSpinner size="2rem" className="mx-auto my-4" />
            ) : (
              !!availableWasteBottles.length && (
                <div className="grid grid-cols-3 gap-2">
                  <InputCheckbox
                    name="selectAll"
                    value={
                      Object.values(form.wasteBottleIds).filter((v) => v)
                        .length === availableWasteBottles.length
                        ? ['active']
                        : []
                    }
                    checkOptionValue="active"
                    label="Seleccionar todas"
                    handleChange={handleSelectAllBottles}
                    className="col-span-3"
                  />
                  {availableWasteBottles.map((bottle) => (
                    <InputCheckbox
                      key={bottle.id}
                      name={`wasteBottleIds.${bottle.id}`}
                      value={form.wasteBottleIds[bottle.id] ? ['active'] : []}
                      checkOptionValue="active"
                      label={`Botella #${bottle.id} ${
                        bottle.trackingCode ? `(${bottle.trackingCode})` : ''
                      }`}
                      handleChange={(name, value) =>
                        handleChange(name, !!value.includes('active'))
                      }
                    />
                  ))}
                </div>
              )
            )}
          </div>
        </div>
        <Button
          label={isEditing ? 'Actualizar' : 'Crear'}
          disabled={!submitEnabled}
          className="ml-auto mt-4 sticky bottom-0 min-w-[10rem]"
          handleClick={handleSubmit}
        />
      </form>
    </Modal>
  );
};

export default BatchFormModal;
