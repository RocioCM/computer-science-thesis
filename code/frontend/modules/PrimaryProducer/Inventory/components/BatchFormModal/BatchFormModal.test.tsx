import { render, screen, fireEvent, act } from '@testing-library/react';
import BatchFormModal from './BatchFormModal';

describe('Primary Producer Inventory - BatchFormModal', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    const modalRoot = document.createElement('div');
    modalRoot.setAttribute('id', 'modal-root');
    document.body.appendChild(modalRoot);
  });

  afterEach(() => {
    const modalRoot = document.getElementById('modal-root');
    if (modalRoot) {
      document.body.removeChild(modalRoot);
    }
  });

  it('Matches snapshot', () => {
    const { container } = render(
      <BatchFormModal
        editingId={1}
        handleCancel={() => {}}
        handleSuccess={() => {}}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
