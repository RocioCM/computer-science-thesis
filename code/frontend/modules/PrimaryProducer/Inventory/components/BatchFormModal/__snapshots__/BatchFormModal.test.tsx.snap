// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Primary Producer Inventory - BatchFormModal Matches snapshot 1`] = `
<div>
  <div
    id="modal-root"
  >
    <div
      class="bg-n10 bg-opacity-70 py-2xl px-xl w-full h-screen fixed top-0 left-0 z-modal flex justify-center items-center animate__animated animate__fadeIn"
      data-testid="modal"
    >
      <div
        class="bg-n0 min-w-[300px] min-h-[200px] rounded-rm overflow-y-auto flex flex-col items-center relative cursor-default animate__animated animate__zoomIn w-[700px] h-[700px] max-h-full max-w-full px-xl py-xl"
      >
        <div
          class="flex justify-end w-full h-0 sticky top-0 z-10"
        >
          <img
            alt="Cerrar"
            class="h-6 w-auto cursor-pointer object-contain"
            data-nimg="1"
            decoding="async"
            height="30"
            loading="lazy"
            src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
            srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
            style="color: transparent;"
            width="30"
          />
        </div>
        <h2
          class="w-full bg-n0"
        >
          Editar
           lote de botellas
           
          #1
        </h2>
        <form
          class="h-full w-full overflow-auto hide-scroll relative pt-2"
        >
          <div
            class="flex flex-col gap-2"
          >
            <div
              class="grid grid-cols-2 gap-2"
            >
              <div
                class="w-full"
              >
                <span
                  class="block relative z-10 w-max transition-all text-p mb-xs top-0 left-0 text-n10 font-semibold"
                >
                  Cantidad de envases
                </span>
                <div
                  class="relative flex flex-row flex-nowrap justify-between items-center w-full"
                >
                  <div
                    class="absolute left-0 h-full w-max flex flex-row items-center gap-xs"
                  />
                  <input
                    class="input block w-full relative border bg-transparent outline-none transition-shadow p rounded-rs border-n3 px-m py-xs placeholder:text-n2 h-10 hover:border-n2 inputFocus focus:border-p1 focus:border-2 disabled:text-n3 disabled:bg-n1 disabled:border-n2 disabled:placeholder:text-transparent"
                    data-testid="input"
                    form="[object Object]"
                    inputlabel="Cantidad de envases"
                    name="quantity"
                    options=""
                    placeholder="Ingresá la cantidad de envases"
                    required=""
                    type="number"
                    value=""
                  />
                  <div
                    class="absolute right-0 h-full w-max flex flex-row items-center gap-xs"
                  />
                </div>
              </div>
              <div
                class="w-full"
              >
                <span
                  class="block relative z-10 w-max transition-all text-p mb-xs top-0 left-0 text-n10 font-semibold"
                >
                  Peso por envase (en gramos)
                </span>
                <div
                  class="relative flex flex-row flex-nowrap justify-between items-center w-full"
                >
                  <div
                    class="absolute left-0 h-full w-max flex flex-row items-center gap-xs"
                  />
                  <input
                    class="input block w-full relative border bg-transparent outline-none transition-shadow p rounded-rs border-n3 px-m py-xs placeholder:text-n2 h-10 hover:border-n2 inputFocus focus:border-p1 focus:border-2 disabled:text-n3 disabled:bg-n1 disabled:border-n2 disabled:placeholder:text-transparent"
                    data-testid="input"
                    form="[object Object]"
                    inputlabel="Peso por envase (en gramos)"
                    name="bottleType.weight"
                    options=""
                    placeholder="Ingresá el peso por envase (gramos)"
                    required=""
                    type="number"
                    value=""
                  />
                  <div
                    class="absolute right-0 h-full w-max flex flex-row items-center gap-xs"
                  />
                </div>
              </div>
              <div
                class="w-full"
              >
                <span
                  class="block relative z-10 w-max transition-all text-p mb-xs cursor-pointer top-0 left-0 text-n10 font-semibold"
                >
                  Color del envase
                </span>
                <button
                  class="relative flex flex-row flex-nowrap justify-between items-center w-full border bg-transparent outline-none transition-shadow gap-xs p text-left rounded-rs border-n3 px-m py-m placeholder:text-n2 h-10 hover:border-n2 dropdonwFocus focus:border-p1 focus:border-2 disabled:text-n3 disabled:bg-n1 disabled:border-n2"
                  data-testid="dropdown"
                  name="bottleType.color"
                >
                  <span
                    class="flex-1 min-h-4"
                  >
                    Seleccioná el color del envase
                  </span>
                  <i
                    class="inline-block w-2.5 h-2.5 rotate-0 "
                    data-testid="chevron-icon"
                  >
                    <svg
                      class="w-full h-full block transition-transform duration-300 ease-out"
                      fill="none"
                      viewBox="0 0 15 9"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M14.7 1.89992L13.5 0.919922L7.50001 6.09992L1.50001 0.919921L0.300013 1.89992L7.50001 8.19992L14.7 1.89992Z"
                        fill="var(--colors-neutrals-n10)"
                      />
                    </svg>
                  </i>
                </button>
              </div>
              <div
                class="w-full"
              >
                <span
                  class="block relative z-10 w-max transition-all text-p mb-xs top-0 left-0 text-n10 font-semibold"
                >
                  Espesor (en milímetros)
                </span>
                <div
                  class="relative flex flex-row flex-nowrap justify-between items-center w-full"
                >
                  <div
                    class="absolute left-0 h-full w-max flex flex-row items-center gap-xs"
                  />
                  <input
                    class="input block w-full relative border bg-transparent outline-none transition-shadow p rounded-rs border-n3 px-m py-xs placeholder:text-n2 h-10 hover:border-n2 inputFocus focus:border-p1 focus:border-2 disabled:text-n3 disabled:bg-n1 disabled:border-n2 disabled:placeholder:text-transparent"
                    data-testid="input"
                    form="[object Object]"
                    inputlabel="Espesor (en milímetros)"
                    name="bottleType.thickness"
                    options=""
                    placeholder="Ingresá el espesor (mm)"
                    required=""
                    type="number"
                    value=""
                  />
                  <div
                    class="absolute right-0 h-full w-max flex flex-row items-center gap-xs"
                  />
                </div>
              </div>
              <div
                class="w-full"
              >
                <span
                  class="block relative z-10 w-max transition-all text-p mb-xs top-0 left-0 text-n10 font-semibold"
                >
                  Tipo/forma del envase
                </span>
                <div
                  class="relative flex flex-row flex-nowrap justify-between items-center w-full"
                >
                  <div
                    class="absolute left-0 h-full w-max flex flex-row items-center gap-xs"
                  />
                  <input
                    class="input block w-full relative border bg-transparent outline-none transition-shadow p rounded-rs border-n3 px-m py-xs placeholder:text-n2 h-10 hover:border-n2 inputFocus focus:border-p1 focus:border-2 disabled:text-n3 disabled:bg-n1 disabled:border-n2 disabled:placeholder:text-transparent"
                    data-testid="input"
                    form="[object Object]"
                    inputlabel="Tipo/forma del envase"
                    name="bottleType.shapeType"
                    options=""
                    placeholder="Ingresá el tipo/forma del envase"
                    required=""
                    type="text"
                    value=""
                  />
                  <div
                    class="absolute right-0 h-full w-max flex flex-row items-center gap-xs"
                  />
                </div>
              </div>
            </div>
            <hr
              class="my-4"
            />
            <div
              class="flex flex-col gap-2"
            >
              <h3>
                Composición
              </h3>
              <div
                class="flex gap-2"
              >
                <div
                  class="w-full"
                >
                  <span
                    class="block relative z-10 w-max transition-all text-p mb-xs top-0 left-0 text-n10 font-semibold"
                  >
                    Nombre del material
                  </span>
                  <div
                    class="relative flex flex-row flex-nowrap justify-between items-center w-full"
                  >
                    <div
                      class="absolute left-0 h-full w-max flex flex-row items-center gap-xs"
                    />
                    <input
                      class="input block w-full relative border bg-transparent outline-none transition-shadow p rounded-rs border-n3 px-m py-xs placeholder:text-n2 h-10 hover:border-n2 inputFocus focus:border-p1 focus:border-2 disabled:text-n3 disabled:bg-n1 disabled:border-n2 disabled:placeholder:text-transparent"
                      data-testid="input"
                      form="[object Object]"
                      inputlabel="Nombre del material"
                      name="bottleType.composition.0.name"
                      options=""
                      placeholder="Ingresá el nombre del material"
                      required=""
                      type="text"
                      value=""
                    />
                    <div
                      class="absolute right-0 h-full w-max flex flex-row items-center gap-xs"
                    />
                  </div>
                </div>
                <div
                  class="w-full"
                >
                  <span
                    class="block relative z-10 w-max transition-all text-p mb-xs top-0 left-0 text-n10 font-semibold"
                  >
                    Cantidad de material
                  </span>
                  <div
                    class="relative flex flex-row flex-nowrap justify-between items-center w-full"
                  >
                    <div
                      class="absolute left-0 h-full w-max flex flex-row items-center gap-xs"
                    />
                    <input
                      class="input block w-full relative border bg-transparent outline-none transition-shadow p rounded-rs border-n3 px-m py-xs placeholder:text-n2 h-10 hover:border-n2 inputFocus focus:border-p1 focus:border-2 disabled:text-n3 disabled:bg-n1 disabled:border-n2 disabled:placeholder:text-transparent"
                      data-testid="input"
                      form="[object Object]"
                      inputlabel="Cantidad de material"
                      name="bottleType.composition.0.amount"
                      options=""
                      placeholder="Ingresá la cantidad de material"
                      required=""
                      type="number"
                      value=""
                    />
                    <div
                      class="absolute right-0 h-full w-max flex flex-row items-center gap-xs"
                    />
                  </div>
                </div>
                <div
                  class="w-full"
                >
                  <span
                    class="block relative z-10 w-max transition-all text-p mb-xs top-0 left-0 text-n10 font-semibold"
                  >
                    Unidad de medida
                  </span>
                  <div
                    class="relative flex flex-row flex-nowrap justify-between items-center w-full"
                  >
                    <div
                      class="absolute left-0 h-full w-max flex flex-row items-center gap-xs"
                    />
                    <input
                      class="input block w-full relative border bg-transparent outline-none transition-shadow p rounded-rs border-n3 px-m py-xs placeholder:text-n2 h-10 hover:border-n2 inputFocus focus:border-p1 focus:border-2 disabled:text-n3 disabled:bg-n1 disabled:border-n2 disabled:placeholder:text-transparent"
                      data-testid="input"
                      form="[object Object]"
                      inputlabel="Unidad de medida"
                      name="bottleType.composition.0.measureUnit"
                      options=""
                      placeholder="Ingresá la unidad de medida de la cantidad de material"
                      required=""
                      type="text"
                      value=""
                    />
                    <div
                      class="absolute right-0 h-full w-max flex flex-row items-center gap-xs"
                    />
                  </div>
                </div>
              </div>
              <button
                class="relative flex items-center justify-center gap-s rounded-rs font-semibold h-10 disabled:cursor-default disabled:bg-opacity-40  focusButton bg-transparent text-p1 border border-p1 py-s px-m w-max"
                data-testid="button"
              >
                <i
                  class="fa-solid fa-plus"
                  data-testid="fa-icon"
                />
                 Agregar material
              </button>
            </div>
            <hr
              class="my-4"
            />
            <div
              class="grid grid-cols-2 gap-2"
            >
              <div
                class="w-full"
              >
                <span
                  class="block relative z-10 w-max transition-all text-p mb-xs top-0 left-0 text-n10 font-semibold"
                >
                  Fecha de producción
                </span>
                <div
                  class="relative flex flex-row flex-nowrap justify-between items-center w-full"
                >
                  <div
                    class="absolute left-0 h-full w-max flex flex-row items-center gap-xs"
                  />
                  <input
                    class="input block w-full relative border bg-transparent outline-none transition-shadow p rounded-rs border-n3 px-m py-xs placeholder:text-n2 h-10 hover:border-n2 inputFocus focus:border-p1 focus:border-2 disabled:text-n3 disabled:bg-n1 disabled:border-n2 disabled:placeholder:text-transparent"
                    data-testid="input"
                    form="[object Object]"
                    inputlabel="Fecha de producción"
                    max="2025-05-03"
                    name="createdAt"
                    options=""
                    placeholder="Seleccioná la fecha de producción"
                    required=""
                    type="date"
                    value="2025-05-03"
                  />
                  <div
                    class="absolute right-0 h-full w-max flex flex-row items-center gap-xs"
                  />
                </div>
              </div>
              <div
                class="w-full"
              >
                <span
                  class="block relative z-10 w-max transition-all text-p mb-xs top-0 left-0 text-n10 font-semibold"
                >
                  Localización de producción
                </span>
                <div
                  class="relative flex flex-row flex-nowrap justify-between items-center w-full"
                >
                  <div
                    class="absolute left-0 h-full w-max flex flex-row items-center gap-xs"
                  />
                  <input
                    class="input block w-full relative border bg-transparent outline-none transition-shadow p rounded-rs border-n3 px-m py-xs placeholder:text-n2 h-10 hover:border-n2 inputFocus focus:border-p1 focus:border-2 disabled:text-n3 disabled:bg-n1 disabled:border-n2 disabled:placeholder:text-transparent"
                    data-testid="input"
                    form="[object Object]"
                    inputlabel="Localización de producción"
                    name="bottleType.originLocation"
                    options=""
                    placeholder="Ingresá la localización de producción"
                    required=""
                    type="text"
                    value=""
                  />
                  <div
                    class="absolute right-0 h-full w-max flex flex-row items-center gap-xs"
                  />
                </div>
              </div>
              <label
                class="w-full block col-span-2"
              >
                <span
                  class="block relative z-10 w-max transition-all text-p mb-xs top-0 left-0 text-n10 font-semibold"
                >
                  Observaciones adicionales (opcional)
                </span>
                <textarea
                  class="input block w-full h-max relative resize-none border bg-transparent outline-none transition-shadow p rounded-rs border-n3 px-m py-m placeholder:text-n2 min-h-[5rem] hover:border-n2 inputFocus focus:border-p1 focus:border-2 disabled:text-n3 disabled:bg-n1 disabled:border-n2 disabled:placeholder:text-transparent"
                  form="[object Object]"
                  inputlabel="Observaciones adicionales (opcional)"
                  name="bottleType.extraInfo"
                  options=""
                  placeholder="Ingresá observaciones adicionales"
                  type="textarea"
                />
              </label>
            </div>
          </div>
          <button
            class="relative flex items-center justify-center gap-s rounded-rs font-semibold h-10 disabled:cursor-default disabled:bg-opacity-40  focusButton bg-p1 text-n0 py-s px-m w-max ml-auto mt-4 sticky bottom-0 min-w-[10rem]"
            data-testid="button"
            disabled=""
          >
            <span>
              Actualizar
            </span>
          </button>
        </form>
      </div>
      <style>
        
          body {
            overflow: hidden;
          }
        
      </style>
    </div>
  </div>
</div>
`;
